lib/data_logger/destination.ex

```diff
@@ -37,28 +40,40 @@ defmodule DataLogger.Destination do
   """
 
   @type prefix :: atom() | String.t()
-  @type(send_result :: :ok | {:ok, result :: term()}, {:error, reason :: term()})
+  @type options :: map()
+  @type send_result ::
+          :ok
+          | {:ok, result :: term()}
+          | {:error, reason :: term()}
+          | {:ok, result :: term(), options()}
+          | {:error, reason :: term(), options()}
 
-  @callback send_data(prefix(), data :: term(), options :: keyword()) :: send_result()
+  @callback send_data(prefix(), data :: term(), options()) :: send_result()
 
-  @callback on_error(error :: term(), prefix(), data :: term(), options :: keyword()) :: :ok
-  @callback on_success(result :: term(), prefix(), data :: term(), options :: keyword()) :: :ok
+  @callback initialize(options()) :: options()
 
-  @optional_callbacks on_error: 4,
+  @callback on_error(error :: term(), prefix(), data :: term(), options()) :: :ok
+  @callback on_success(result :: term(), prefix(), data :: term(), options()) :: :ok
+
+  @optional_callbacks initialize: 1,
+                      on_error: 4,
                       on_success: 4
 
   @doc false
   defmacro __using__(_) do
     quote do
       @behaviour DataLogger.Destination
 
+      @doc false
+      def initialize(options), do: options
+
       @doc false
       def on_error(_, _, _, _), do: :ok
 
       @doc false
       def on_success(_, _, _, _), do: :ok
 
-      defoverridable on_error: 4, on_success: 4
+      defoverridable initialize: 1, on_error: 4, on_success: 4
```

Comment:
If all callbacks have a default implementation and are overridable this is the same as `defoverridable DataLogger.Destination`

Reply:
Not all of them are. There is one which is mandatory.